content/blog/2022-05-05-e2e-cv-pt3.md

```diff
@@ -0,0 +1,232 @@
+      # Create CML report
+      echo "## Metrics" >> report.md
+      pipenv run dvc metrics show --md >> report.md
+      echo "## A few random test images" >> report.md
+      for file in $(ls data/test_preds/ |sort -R |tail -20); do
+        cml publish data/test_preds/$file --md >> report.md
+      done
+      cml send-comment --pr --update report.md
+      ...
+      ```
+
+      The report structure is fully customizable. Below is an example of what
+      the PR and the CML report would look like in this case:
+
+      ![PR and CML report](/uploads/images/2022-05-05/pr_cml_report.png '=800')
+
+   At this point, we can assess the results in Iterative Studio and GitHub and
+   decide whether we want to accept the PR or keep experimenting.
+
+2. [Workflow for deploying to the development environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/2-develop.yaml)
+
+   ![Workflow for deploying to the development environment](/uploads/images/2022-05-05/workflow_dev.png '=800')
+
+   Once we are happy with our model's performance on the experiment branch, we
+   can merge it into the development branch. This would trigger a different
+   CI/CD job that will:
+
+   1. retrain the model with the new parameters (as the `dev` branch might
+      contain changes not present in the `exp` branch). This step looks almost
+      identical to the third step in the above workflow.
+   2. deploy the web REST API application (that incorporates the new model) to a
+      development endpoint on Heroku:
+
+      ```yaml
+      ---
+      deploy-dev-api:
+        needs: train-and-push
+        runs-on: ubuntu-latest
+        steps:
+          - uses: actions/checkout@v2
+          - uses: actions/download-artifact@master
+            with:
+              name: model_pickle
+              path: models
+          - uses: akhileshns/heroku-deploy@v3.12.12
+            with:
+              heroku_api_key: ${{secrets.HEROKU_API_KEY}}
+              heroku_app_name: 'demo-api-mag-tiles-dev'
+              heroku_email: 'alexkim@iterative.ai'
+              team: 'iterative-sandbox'
+              usedocker: true
+      ```
+
+      Now we can test our API and assess the end-to-end performance of the
+      overall solution.
+
+3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/3-deploy.yaml)
+
+   ![Workflow for deploying to the production environment](/uploads/images/2022-05-05/workflow_prod.png '=800')
+
+   If we've thoroughly tested and monitored the performance of our development
+   web API, we can merge the development branch into the main branch of our
+   repository. Again, this triggers the third CI/CD workflow that deploys the
+   code from the main branch to the production API. This looks identical to the
+   deployment into the development environment except for `heroku_app_name`
+   which is now set to `"demo-api-mag-tiles-prod"`.
```

Comment:
The last steps read too abstract to me. I understand that it's not focusing on our tools but feels like we are closing the series and I don't know where the model is actually deployed or how to use it.

Would love a simple diagram or image showing how to use the API and get the prediction pack, maybe an extension of the one used at the beginning of part 1? A screenshot of how the API looks in heroku?

Reply:
Yup, I added examples in both `curl` and `python` with input and output images.